### PR TITLE
Adding a way to get thrown errors that have been forgotten about

### PR DESCRIPTION
**Pull request by fresheneesz**
If you throw an error into a future, but you never wait or detach that future, those throws would be lost. This commit adds a way to grab all the forgotten about errors and do something about them (probably log them so you can fix the fact that you forgot about them). 

I also made some minor improvements, modularized some code duplication in Future.wait, and added a todo for detach to cause errors to be thrown from the future.throw call _always_ instead of sometimes throwing it from the future.detach call itself.

### CODE REVIEW

**Comment by fresheneesz** [comment]
@laverdet I just read something in the docs for bluebird promise. See the stuff under the heading "error handling" https://github.com/petkaantonov/bluebird . 

> download().then(...).then(...).done();

>For handling this problem, in my opinion, this is completely unacceptable and pointless. The user must remember to explicitly call .done and that cannot be justified when the problem is forgetting to create an error handler in the first place.

>The second approach, which is what bluebird by default takes, is to call a registered handler if a rejection is unhandled by the start of a second turn. The default handler is to write the stack trace to stderr or console.error in browsers. This is close to what happens with synchronous code - your code doesn't work as expected and you open console and see a stack trace. Nice.

I like that approach - the default should be that something is waiting on that future when the future is resolved (or at least something has started waiting on it in the same continuation as where the future was created before that continuation ends). I think this technique could work with fibers futures much better than the proposal I originally made.

By default, an error would be thrown if an error comes through on a future that has not been waited on one cycle after the error comes through. In cases where you're getting fancy and deferring waiting on some futures, you can explicitly call something on them to turn off that behavior. Examples:

```
var a = new Future
a.throw(new Error("whatever") // after one cycle, this error is thrown into the scheduler

var b = new Future
b.throw(new Error("2"))
b.afar()  // dunno what a good word would be here - detach would be a good word here, so would defer, but all those have overriden meanings already

// for b, no error is thrown to the scheduler, and the code patiently waits forever for something to use that future
```

In the error thrown to the scheduler in the default case, it can have a nice note about looking at the docs for the 'afar' method (or whatever you'd want to call it), so that a change can be made at that point. 

In short, this solves the problem of lost exceptions pretty nicely without the possibility of memory leaks, and with a clear solution to the issue you mentioned about knowing "how many pending futures is 'ok'", since the `afar` method would allow explicit definition of which futures are ok to forget about as far as unhandled exceptions.

The only case this could still involve losing an exception is if you stopped using the future, but forgot to remove the call to 'afar' - which is *far* less likely than simply not waiting on a future (pun.. retroactively intended).
